processor/detector.go

```diff
@@ -0,0 +1,33 @@
+package processor
+
+import (
+	"errors"
+	"strings"
+)
+
+// Given some content attempt to determine if it has a #! that maps to a known language and return the language
+func DetectSheBang(content string) (string, error) {
+	if !strings.HasPrefix(content, "#!") {
+		return "", errors.New("Missing #!")
+	}
+
+	index := strings.Index(content, "\n")
+
+	if index != -1 {
+		content = content[:index]
+	}
+
+	cln := strings.Replace(content, " ", "", -1)
+	cln = strings.Replace(cln, "-w", "", -1)
+
+	for k, v := range ShebangLookup {
+		for _, x := range v {
+			// detects both full path and env usage
+			if strings.HasSuffix(cln, "/"+x) || strings.Contains(content, " "+x) {
+				return k, nil
+			}
+		}
+	}
```

Comment:
Replacing `-w` isn't really appropriate as it's a perl specific argument, not a general shebang thing. The argument after the executable name is completely arbitrary and gets interpreted by the executable.

I think it's worth extracting the executable/arguments out properly, rather than guessing based on substrings.

This should pull both out with a single pass through the string:

```go
func scanShebang(shebang string) (string, string, bool) {
    if !strings.HasPrefix(shebang, "#!") {
        return "", "", false
    }

    slashIndex := 0
    wsStartIndex := 0
    wsEndIndex := 0

    body := strings.TrimSpace(shebang[2:])
    for i := 0; i <  len(body); i++ {
        if body[i] == '/' {
            slashIndex = i
        }

        if body[i] == ' ' || body[i] == '\t' {
            if wsStartIndex == 0 {
                wsStartIndex = i
            }

            wsEndIndex = i
        }

        if wsEndIndex != 0 && wsEndIndex != i {
            break
        }

        if body[i] == '\n' {
            break
        }
    }

    exe := body[slashIndex+1:wsStartIndex]

    arg := ""
    if wsEndIndex != 0 {
        arg = body[wsEndIndex+1:]
    }

    return exe, arg, true
}
```

Reply:
The above fails on a few of my test cases sadly, due to string slicing with wsStartIndex being at 0. I was modifying it to work then realised.

There may be an easier approach. There are only 2 cases that actually matter.

1. if `/env ` is present then the next non-blank string to either a space, newline or tab is the program
2. if `/env ` is not present then the program is whatever follows the last occurrence of `/` to a space, newline or tab.

There are no other cases. The above accounts for the 3 possible cases,

1. `#!/usr/bin/perl -w`
2. `#!/usr/bin/perl`
3. `#!/usr/bin/env perl -w`

It does not matter how many spaces follow the env or if any follow the command.

Ill have a think on this. I have a feeling there is a simple solution here that works for all cases.

Reply:
You're right, I didn't properly handle the case of not having an argument when I slapped this together. A condition before/around the `exe := body[slashIndex+1:wsStartIndex]` should be all it needs to fix it though.

> There may be an easier approach. There are only 2 cases that actually matter.

The problem is that you need to match on whitespace, which means that the various `strings.*` functions aren't going to cut it. You could do it with a regular expression, but most people would find `^#![^ \t]*/([^ \t]+)([ \t]+(.*))?$` illegible. Hell, I'm generally pretty good with regular expressions and I think I might struggle with that one if I hadn't just written it.

> 3. `#!/usr/bin/env perl -w`

This case is actually invalid, since shebang parsing doesn't split arguments. Looks like GNU `env` does accept `#!/usr/bin/env --split-string=perl -w`, but I've never seen that form in the wild.

Reply:
No I don't mean using regex or string. Another small state machine should do it. Can do the #! check in there as well.

You start in nothing, move to shebang, then to first / where you record the last found one till you hit a space or newline, and as such know exactly where to make the cuts. Very similar to what you did in fact.

Reply:
Committed something that is currently broken to demonstrate how it could work. Works in this case for one case at least.

Reply:
Oh, I see what you mean. I thought that trying to embed the `env` logic into the state machine would overcomplicate it. But it's a valid solution if you want to go that way.

In your demo sample, `state == 0` and `state == 1` are the same thing. I'm pretty sure the only separate states you need are "before whitespace" and "after whitespace".

Reply:
Yep, just a quick thing I threw together. 100% not ready for release till I actually fix it :)

Reply:
Have a look at whats in there now @dbaggerman moved to use a state machine. 

Benchmarks look promising and it works with all existing cases

```
$ go test -run=SheBang -bench=.                                                                                                                                                                                     goos: linux                                                                                                                                                                                                         goarch: amd64                                                                                                                                                                                                       pkg: github.com/boyter/scc/processor                                                                                                                                                                                BenchmarkScanSheBangFuzz-8                                                500756              2329 ns/op
```